pkg/skaffold/runner/label/label.go

```diff
@@ -86,67 +90,81 @@ func updateRuntimeObject(client clientgo.Interface, labels map[string]string, re
 	}
 	var err error
 	obj := *res.Obj
+	originalJSON, _ := json.Marshal(obj)
+	modifiedObj := obj.DeepCopyObject()
 	switch obj.(type) {
 	case *corev1.Pod:
-		p := obj.(*corev1.Pod)
-		addSkaffoldLabels(labels, &p.ObjectMeta)
-		pods := client.CoreV1().Pods(retrieveNamespace(res.Namespace, p.ObjectMeta))
-		_, err = pods.UpdateStatus(p)
+		apiObject := modifiedObj.(*corev1.Pod)
+		addSkaffoldLabels(labels, &apiObject.ObjectMeta)
+		modifiedJSON, _ := json.Marshal(modifiedObj)
+		p, _ := patch.CreateTwoWayMergePatch(originalJSON, modifiedJSON, apiObject)
+		_, err = client.CoreV1().Pods(retrieveNamespace(res.Namespace, apiObject.ObjectMeta)).Patch(apiObject.Name, types.StrategicMergePatchType, p)
 	case *appsv1.Deployment:
-		d := obj.(*appsv1.Deployment)
-		addSkaffoldLabels(labels, &d.ObjectMeta)
-		deployments := client.AppsV1().Deployments(retrieveNamespace(res.Namespace, d.ObjectMeta))
-		_, err = deployments.Update(d)
+		apiObject := modifiedObj.(*appsv1.Deployment)
+		addSkaffoldLabels(labels, &apiObject.ObjectMeta)
+		modifiedJSON, _ := json.Marshal(modifiedObj)
+		p, _ := patch.CreateTwoWayMergePatch(originalJSON, modifiedJSON, apiObject)
+		_, err = client.AppsV1().Deployments(retrieveNamespace(res.Namespace, apiObject.ObjectMeta)).Patch(apiObject.Name, types.StrategicMergePatchType, p)
 	case *appsv1beta1.Deployment:
-		d := obj.(*appsv1beta1.Deployment)
-		addSkaffoldLabels(labels, &d.ObjectMeta)
-		deployments := client.AppsV1beta1().Deployments(retrieveNamespace(res.Namespace, d.ObjectMeta))
-		_, err = deployments.Update(d)
+		apiObject := modifiedObj.(*appsv1beta1.Deployment)
+		addSkaffoldLabels(labels, &apiObject.ObjectMeta)
+		modifiedJSON, _ := json.Marshal(modifiedObj)
+		p, _ := patch.CreateTwoWayMergePatch(originalJSON, modifiedJSON, apiObject)
+		_, err = client.AppsV1beta1().Deployments(retrieveNamespace(res.Namespace, apiObject.ObjectMeta)).Patch(apiObject.Name, types.StrategicMergePatchType, p)
 	case *appsv1beta2.Deployment:
-		d := obj.(*appsv1beta2.Deployment)
-		addSkaffoldLabels(labels, &d.ObjectMeta)
-		deployments := client.AppsV1beta2().Deployments(retrieveNamespace(res.Namespace, d.ObjectMeta))
-		_, err = deployments.Update(d)
+		apiObject := modifiedObj.(*appsv1beta2.Deployment)
```

Comment:
At this point it might make sense to reduce this switch statement to lower the boilerplate. One idea is a map (or two maps) of type (appsv1beta2.Deployment) -> functions.

One function could be "apiObjectGetter", which does the type conversion, and the other could do the Patch itself.

Then this switch statement is a lookup of type to functions, then calls those functions. WDYT?

Reply:
yeah, I agree I should have addressed this before. just refactored it, I'll detail in the top-level comment